### DISCUSSION TITLE
Json field response parsing error

### DISCUSSION THREAD

**Discussion by ljbuza** [fastapi/fastapi#8480] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from typing import List
from pydantic import BaseModel, Json

app = FastAPI()

# schema
class JobResponse(BaseModel):
    name: str
    input: Json


# main route
@app.get("/jobs", response_model=List[JobResponse])
def get_jobs():
    return [JobResponse(name="foo", input='{"foo": "bar"}')]
```

### Description

Hitting the `/jobs` endpoint throws:

```python
  File "...python3.8/site-packages/fastapi/routing.py", line 235, in app
    response_data = await serialize_response(
  File "...python3.8/site-packages/fastapi/routing.py", line 138, in serialize_response
    raise ValidationError(errors, field.type_)
pydantic.error_wrappers.ValidationError: 1 validation error for JobResponse
response -> 0 -> input
  JSON object must be str, bytes or bytearray (type=type_error.json)
```
However, creating the same JobResponse object outside of fastapi works just fine:

```python
In[1]: JobResponse(name="foo", input='{"bar": "baz"}')
Out[1]: JobResponse(name='foo', input={'bar': 'baz'})
```

### Operating System

Linux

### FastAPI Version

0.75.2

### Python Version

3.8

**Comment by jtc42** [comment]
I think the issue is that you're basically running the Pydantic validation step twice. Once when you create your JobResponse objects in your return statement, then AGAIN when response_model triggers the validation. 

If you replace your list of JobResponse objects with a list of dictionaries for example, containing the same data, then exclusively let response_model run the validation, does that work?

E.g.

```python
from fastapi import FastAPI
from typing import List
from pydantic import BaseModel, Json

app = FastAPI()

# schema
class JobResponse(BaseModel):
    name: str
    input: Json


# main route
@app.get("/jobs", response_model=List[JobResponse])
def get_jobs():
    return [{name: "foo", input: '{"foo": "bar"}'}]
```

If the Pydantic Json validation step converts your string into a new type, which the Json type cannot itself accept as an input, then it would trigger the exception you're seeing. The "solution" I'm suggesting basically prevents that happening by only having your `input` argument converted once.

Does this make any kind of sense? I'm struggling to convey what I mean...

**Reply by jonaskyhse** [reply]
I have had the same issue and I can confirm that not using the response_model parameter and returning a dictionary works but not both returning a response_model instance and using the parameter response_model. However, this is something that is only a problem for the Json type which makes me think that it can be considered a bug on the type. Other pydantic models work perfectly fine.

I would say, that this is because there is a side effect in running the validation cause the validation is actually deserialising the json-string at the same time and mutating the instance inlace. Maybe this could be fixed in pydantic?

**Comment by ljbuza** [comment]
I may have presented an oversimplification of the problem for the sake of a concise example,  but it does appear that the problem is that validation is triggering more than once.  The issue only seems to come up with a paginated orm response, illustrated here:

https://github.com/uriyyo/fastapi-pagination/issues/316

With the combination of sqlalchemy, fastapi, pydantic, and fastapi-pagination all in the mix, I'm struggling to debug this.

**Comment by jtc42** [comment]
So the reason I replied is because I think I've faced this exact issue before, but it was a while ago and it's a bit hazy. All I *really* remember was that the issue was related to the validation triggering more than once, and the first time the validation runs, it converts the JSON string into a JSON type, which the `json` module is unable to serialise.

I'm assuming you can't just use a nested Pydantic model instead of a `JSON`-typed field?

The only other thing I can think to suggest is to create either a custom JSON serialiser, or a custom JSON field type that's able to handle being given a JSON-typed object as an input? All of these feel like hacks though...

**Comment by ljbuza** [comment]
In reference to the code in https://github.com/uriyyo/fastapi-pagination/issues/316

changing `input=Json` to the `Input` model results in:

```python
items -> 0 -> input
  value is not a valid dict (type=type_error.dict)
```

changing it to `input=str` doesn't throw errors, but the response is
```python
{"name": "test", "input": "{\"type\": \"foo\", \"val\": 2}"}
```

**Comment by jtc42** [comment]
> In reference to the code in [uriyyo/fastapi-pagination#316](https://github.com/uriyyo/fastapi-pagination/issues/316)
> 
> changing `input=Json` to the `Input` model results in:
> 
> ```python
> items -> 0 -> input
>   value is not a valid dict (type=type_error.dict)
> ```
> 
> changing it to `input=str` doesn't throw errors, but the response is
> 
> ```python
> {"name": "test", "input": "{\"type\": \"foo\", \"val\": 2}"}
> ```

Oh yeah course, sorry the model won't work if you're passing `input` as an actual string within a JSON document. I had assumed it was just a nested bit of JSON within the JSON document (if that makes sense?)

Did you try returning dictionaries instead of `JobResponse` objects, then letting the response_model (or paginator-wrapped response model) do all the type conversions just once?

I appreciate it may not be a good workaround but might help identify a good way around this.

**Comment by ljbuza** [comment]
```python
@app.get("/jobs_pg", response_model=Page[JobResponse])
async def get_jobs_pg(db: Session = Depends(get_db)):
    return paginate(db_get_jobs(db=db))
```

`db_get_jobs(db=db)` returns a sqlalchemy orm response object, with the input field defined as a string.  Now changing `db_get_jobs` to unpack the response objects will "work", but obviously not a very good solution:

```python
def db_get_jobs(db: Session):
    data = db.query(JobTable).all()
    resp = []
    for job in data:
         resp.append({"name": job.name, "input": json.dumps(job.input)})
    return resp
```
I see that within the response fastapi.routing.serialize_response() tries to validate, "input" has already been converted from json to a dict:  

```python
> ~.virtualenvs/bec/lib/python3.8/site-packages/fastapi/routing.py(128)serialize_response()
(Pdb) response_content
{'items': [{'name': 'test', 'input': {'type': 'BEC', 'val': 2}}, {'name': 'string', 'input': {'type': 'BEC', 'val': 22}}], 'total': 2, 'page': 1, 'size': 100}
```
which throws the error:

```python
ErrorWrapper(exc=ValidationError(model='Page[JobResponse]', errors=[{'loc': ('items', 0, 'input'), 'msg': 'JSON object must be str, bytes or bytearray', 'type': 'type_error.json'}, {'loc': ('items', 1, 'input'), 'msg': 'JSON object must be str, bytes or bytearray', 'type': 'type_error.json'}]), loc=('response',))
```

**Reply by xingqiwu55555** [reply]
hi, I met same the problem, any solution for it?

**Comment by rootkit7628** [comment]
Got the same problem.
Did someone found a solution or a workaround for this without removing the `response_model` ? 
Why do even the Pydantic validation throw errors on the `response_model` but not when serializing it on the return.

**Comment by YuriiMotov** [comment, marked as answer]
Works well in current version (0.115.12)

```python
from typing import List

from fastapi import FastAPI
from fastapi.testclient import TestClient
from pydantic import BaseModel, Json

app = FastAPI()


# schema
class JobResponse(BaseModel):
    name: str
    input: Json


# main route
@app.get("/jobs", response_model=List[JobResponse])
def get_jobs():
    return [JobResponse(name="foo", input='{"foo": "bar"}')]


def test_():
    client = TestClient(app)
    resp = client.get("/jobs")
    assert resp.status_code == 200
    assert resp.json() == [{"name": "foo", "input": {"foo": "bar"}}]
```
